### ISSUE TITLE
Investigate filter-media task taking all resources

### ISSUE THREAD

**Issue by alawvt** [VTUL/vtechworks#637]
filter-media is in the crontab. It started up this morning and hogged all resources. We killed it finish the indexing.

I believe there are some new filters in DSpace 6, which may be running on content such as PowerPoints, but we need to check this.

sample
```
java -Xmx2000m -Dfile.encoding=UTF-8 -classpath /dspace/webapps/oai/WEB-INF/classes/:/dspace/lib/JUnitParams-1.0.2.jar:/dspace/lib/activation-1.1.1.jar:/dspace/lib/additions-6.3.jar:/dspace/lib/ant-1.9.1.jar:/dspace/lib/ant-contrib-1.0b3.jar:/dspace/lib/ant-launcher-1.9.1.jar:/dspace/lib/antlr-2.7.7.jar:/dspace/lib/antlr-runtime-3.5.jar:/dspace/lib/aopalliance-1.0.jar:/dspace/lib/aopalliance-repackaged-2.4.0-b31.jar:/dspace/lib/apache-mime4j-core-0.7.2.jar:/dspace/lib/asm-4.1.jar:/dspace/lib/asm-analysis-4.1.jar:/dspace/lib/asm-commons-4.1.jar:/dspace/lib/asm-tree-4.1.jar:/dspace/lib/asm-util-4.1.jar:/dspace/lib/aws-java-sdk-core-1.10.50.jar:/dspace/lib/aws-java-sdk-kms-1.10.50.jar:/dspace/lib/aws-java-sdk-s3-1.10.50.jar:/dspace/lib/axiom-api-1.2.14.jar:/dspace/lib/axiom-impl-1.2.14.jar:/dspace/lib/bcmail-jdk15-1.46.jar:/dspace/lib/bcprov-jdk15-1.46.jar:/dspace/lib/bte-core-0.9.3.5.jar:/dspace/lib/bte-io-0.9.3.5.jar:/dspace/lib/builder-commons-1.0.2.jar:/dspace/lib/commons-beanutils-1.9.2.jar:/dspace/lib/commons-cli-1.3.1.jar:/dspace/lib/commons-codec-1.10.jar:/dspace/lib/commons-collections-3.2.2.jar:/dspace/lib/commons-collections4-4.1.jar:/dspace/lib/commons-configuration-1.10.jar:/dspace/lib/commons-csv-1.0.jar:/dspace/lib/commons-dbcp2-2.1.1.jar:/dspace/lib/commons-digester-1.8.1.jar:/dspace/lib/commons-fileupload-1.3.3.jar:/dspace/lib/commons-io-2.4.jar:/dspace/lib/commons-lang-2.6.jar:/dspace/lib/commons-lang3-3.3.2.jar:/dspace/lib/commons-logging-1.2.jar:/dspace/lib/commons-pool2-2.4.2.jar:/dspace/lib/commons-validator-1.5.0.jar:/dspace/lib/coverity-escapers-1.1.1.jar:/dspace/lib/curvesapi-1.04.jar:/dspace/lib/dnsjava-2.1.7.jar:/dspace/lib/dom4j-1.6.1.jar:/dspace/lib/dspace-api-6.3.jar:/dspace/lib/dspace-api-lang-6.0.6.jar:/dspace/lib/dspace-oai-6.3-classes.jar:/dspace/lib/dspace-services-6.3.jar:/dspace/lib/dspace-tm-extractors-1.0.1.jar:/dspace/lib/ehcache-core-2.4.3.jar:/dspace/lib/elasticsearch-1.4.0.jar:/dspace/lib/flyway-core-4.0.3.jar:/dspace/lib/fontbox-2.0.2.jar:/dspace/lib/geoip2-2.11.0.jar:/dspace/lib/geronimo-activation_1.1_spec-1.1.jar:/dspace/lib/geronimo-javamail_1.4_spec-1.7.1.jar:/dspace/lib/geronimo-stax-api_1.0_spec-1.0.1.jar:/dspace/lib/google-api-client-1.23.0.jar:/dspace/lib/google-api-services-analytics-v3-rev145-1.23.0.jar:/dspace/lib/google-http-client-1.23.0.jar:/dspace/lib/google-http-client-jackson2-1.23.0.jar:/dspace/lib/google-oauth-client-1.23.0.jar:/dspace/lib/gson-2.6.1.jar:/dspace/lib/guava-19.0.jar:/dspace/lib/guava-jdk5-17.0.jar:/dspace/lib/hamcrest-all-1.3.jar:/dspace/lib/handle-6.2.jar:/dspace/lib/hibernate-commons-annotations-4.0.2.Final.jar:/dspace/lib/hibernate-core-4.2.21.Final.jar:/dspace/lib/hibernate-ehcache-4.2.21.Final.jar:/dspace/lib/hibernate-jpa-2.0-api-1.0.1.Final.jar:/dspace/lib/hk2-api-2.4.0-b31.jar:/dspace/lib/hk2-locator-2.4.0-b31.jar:/dspace/lib/hk2-utils-2.4.0-b31.jar:/dspace/lib/httpclient-4.5.1.jar:/dspace/lib/httpclient-cache-4.2.6.jar:/dspace/lib/httpcore-4.4.4.jar:/dspace/lib/httpmime-4.3.1.jar:/dspace/lib/icu4j-56.1.jar:/dspace/lib/im4java-1.4.0.jar:/dspace/lib/jackson-annotations-2.8.11.jar:/dspace/lib/jackson-core-2.8.11.jar:/dspace/lib/jackson-core-asl-1.9.13.jar:/dspace/lib/jackson-databind-2.8.11.1.jar:/dspace/lib/jackson-mapper-asl-1.9.13.jar:/dspace/lib/jargon-1.4.25.jar:/dspace/lib/javassist-3.18.1-GA.jar:/dspace/lib/javax.annotation-api-1.2.jar:/dspace/lib/javax.inject-1.jar:/dspace/lib/javax.inject-2.4.0-b31.jar:/dspace/lib/javax.ws.rs-api-2.0.1.jar:/dspace/lib/jaxb-impl-2.2.5.jar:/dspace/lib/jaxb2-basics-runtime-0.9.5.jar:/dspace/lib/jaxen-1.1.6.jar:/dspace/lib/jbibtex-1.0.10.jar:/dspace/lib/jboss-logging-3.1.0.GA.jar:/dspace/lib/jboss-transaction-api_1.1_spec-1.0.1.Final.jar:/dspace/lib/jcl-over-slf4j-1.7.14.jar:/dspace/lib/jdbm-1.0.jar:/dspace/lib/jdom-1.0.jar:/dspace/lib/jdom-1.1.3.jar:/dspace/lib/jena-arq-2.13.0.jar:/dspace/lib/jena-core-2.13.0.jar:/dspace/lib/jena-iri-1.1.2.jar:/dspace/lib/jena-tdb-1.1.2.jar:/dspace/lib/jersey-client-2.22.1.jar:/dspace/lib/jersey-common-2.22.1.jar:/dspace/lib/jersey-guava-2.22.1.jar:/dspace/lib/joda-time-2.9.2.jar:/dspace/lib/json-20180130.jar:/dspace/lib/jsonld-java-0.5.1.jar:/dspace/lib/jtwig-core-2.0.1.jar:/dspace/lib/jtwig-functions-2.0.1.jar:/dspace/lib/jtwig-spring-2.0.1.jar:/dspace/lib/libthrift-0.9.2.jar:/dspace/lib/log4j-1.2.17.jar:/dspace/lib/lucene-analyzers-common-4.10.2.jar:/dspace/lib/lucene-core-4.10.2.jar:/dspace/lib/lucene-grouping-4.10.2.jar:/dspace/lib/lucene-highlighter-4.10.2.jar:/dspace/lib/lucene-join-4.10.2.jar:/dspace/lib/lucene-memory-4.10.2.jar:/dspace/lib/lucene-misc-4.10.2.jar:/dspace/lib/lucene-queries-4.10.2.jar:/dspace/lib/lucene-queryparser-4.10.2.jar:/dspace/lib/lucene-sandbox-4.10.2.jar:/dspace/lib/lucene-spatial-4.10.2.jar:/dspace/lib/lucene-suggest-4.10.2.jar:/dspace/lib/mail-1.4.7.jar:/dspace/lib/maxmind-db-1.2.2.jar:/dspace/lib/mets-1.5.2.jar:/dspace/lib/noggit-0.5.jar:/dspace/lib/oai4j-0.6b1.jar:/dspace/lib/oclc-harvester2-0.1.12.jar:/dspace/lib/opencsv-2.3.jar:/dspace/lib/opensearch-0.1.jar:/dspace/lib/orcid-jaxb-api-2.1.0.jar:/dspace/lib/oro-2.0.8.jar:/dspace/lib/osgi-resource-locator-1.0.1.jar:/dspace/lib/parboiled-core-1.1.6.jar:/dspace/lib/parboiled-java-1.1.6.jar:/dspace/lib/pdfbox-2.0.2.jar:/dspace/lib/poi-3.17.jar:/dspace/lib/poi-ooxml-3.17.jar:/dspace/lib/poi-ooxml-schemas-3.17.jar:/dspace/lib/poi-scratchpad-3.17.jar:/dspace/lib/postgresql-42.2.1.jar:/dspace/lib/reflections-0.9.9-RC1.jar:/dspace/lib/rome-1.0.jar:/dspace/lib/rome-modules-1.0.jar:/dspace/lib/servlet-api-2.5.jar:/dspace/lib/slf4j-api-1.7.14.jar:/dspace/lib/slf4j-log4j12-1.7.14.jar:/dspace/lib/solr-solrj-4.10.4.jar:/dspace/lib/spatial4j-0.4.1.jar:/dspace/lib/spring-aop-3.2.16.RELEASE.jar:/dspace/lib/spring-beans-3.2.16.RELEASE.jar:/dspace/lib/spring-context-3.2.16.RELEASE.jar:/dspace/lib/spring-context-support-3.2.16.RELEASE.jar:/dspace/lib/spring-core-3.2.16.RELEASE.jar:/dspace/lib/spring-expression-3.2.16.RELEASE.jar:/dspace/lib/spring-jdbc-3.2.16.RELEASE.jar:/dspace/lib/spring-orm-3.2.16.RELEASE.jar:/dspace/lib/spring-tx-3.2.16.RELEASE.jar:/dspace/lib/spring-web-3.2.16.RELEASE.jar:/dspace/lib/spring-webmvc-3.2.16.RELEASE.jar:/dspace/lib/stax-api-1.0.1.jar:/dspace/lib/stax2-api-3.1.1.jar:/dspace/lib/test-support-1.0.3.jar:/dspace/lib/woodstox-core-asl-4.1.4.jar:/dspace/lib/wstx-asl-3.2.0.jar:/dspace/lib/xalan-2.7.0.jar:/dspace/lib/xercesImpl-2.11.0.jar:/dspace/lib/xml-apis-1.4.01.jar:/dspace/lib/xmlbeans-2.6.0.jar:/dspace/lib/xmlunit-1.1.jar:/dspace/lib/xoai-3.2.10.jar:/dspace/lib/zookeeper-3.4.6.jar:/dspace/config org.dspace.app.launcher.ScriptLauncher filter-media
```

**Comment by alawvt** [comment]
Re: Memory use of filter-media. The /dspace/bin/dsspace script has a java parameter in it to limit memory usage by these batch processes. It’s currently set very high (2 gigabytes?) , which will be necessary in order to accommodate the batch process from Terry Brady that upgrades the legacy IDs in the solr statistics records to new UUIDs. - @keithgee 

It was not the memory that was the problem, but CPU, it was using a lot of CPU. And possibly IO from the assetstore. The assetstore, since it's mounted using NFS, has higher latency and possibly lower throughput than local storage. That along with an ongoing re-indexing made everything crawl. - @soumikgh 

The filter-media has some new filters for extracting text from items that previously have not been extracted. So it may have been quite busy processing that content. The filters are defined on line 357 of https://github.com/VTUL/vtechworks/blob/vt_6_x_dev/dspace/config/dspace.cfg. - @alawvt 

I think filter-media does essentially the same thing every time it runs - check for bitstreams that need derivatives and creates them. There might be one PDF file or something that’s causing a problem. We could skip it. - @keithgee 

That would be worth investigating whether it is hanging on certain files or whether it’s just high volume. @alawvt 

But, it makes sense that IO is a bottleneck right now. I think that’s why Terry Brady’s re-indexer scripts ran slower on our systems. He benchmarked with an SSD equipped system. I think there’s a lot of small disk reads and writes involved in updating a solr index. - @keithgee 

@soumikgh has commented out filter-media in the crontab temporarily and will start it when the indexing is complete.

**Comment by alawvt** [comment]
I just did a search for .docx and found many items, such as http://hdl.handle.net/10919/70194 that might be useful for testing the new filters. It has never had a text file or thumbnail.

**Comment by keithgee** [comment]
I ran filter-media and it completed successfully. It created 850 new derivative files, including a text extraction on the item that you named above. (It did not create a thumbnail for this item).

I did see the CPU usage for the process go near 600% at times while it was creating the derivative files, but when it is looking through the database for files to filter - most of the run time of the process  - it is usually at about 40%.

I think that, as was speculated by @alawvt and @soumikgh yesterday, the first time it runs after this particular upgrade it is resource intensive because of the media filters for new file formats, and that future runs are less intensive.

** Closing
